### DISCUSSION TITLE
Should the names of components be case in-sensitive in expressions?

### DISCUSSION THREAD

**Discussion by DeadTomGC** [SpiceSharp/SpiceSharpBehavioral#43] [Q&A] [answered]
Should the names of components be case in-sensitive?  In a lot of Spice simulators, they are, so if you're using a netlist generated by some program, you might have mixed case between circuit elements and expressions.  I personally had to deal with this.

**Comment by svenboulanger** [comment, marked as answer]
Hi @DeadTomGC ,

Spice#.Behavioral will (or should) copy the behavior from the circuit and simulation it is working with.

To make the name of entities/components case-insensitive, you have to specify an `IEqualityComparer<string>` when creating the `EntityCollection` that stores the components. Any simulation that is run using this collection will use that comparer.

```csharp
var ec = new EntityCollection(StringComparer.OrdinalIgnoreCase);
var circuit = new Circuit(ec); // Only necessary if you want circuit operations
```

In order to have case-insensitive node-names, you would have to specify a comparer for nodes in the `BiasingParameters` of a simulation. However, after looking into your question I can see that it is in fact not possible.

I took the liberty to open an issue in Spice# [here](https://github.com/SpiceSharp/SpiceSharp/issues/139).

**Reply by svenboulanger** [reply]
As a workaround, you can transform all names, expressions, nodes, etc. to lowercase.

**Reply by DeadTomGC** [reply]
Thanks for the tips! Converting everything to lower was indeed my approach, but I prefer the idea of using a StringComparer.  Thanks for the response! 

I've really had a lot of fun using Spice# so far and It's generally been very easy to work with.   I'm using the mid simulation modification feature to do Hardware in the loop simulations.  So, hardware dictates what certain sources do in the simulation based on a data from the simulation and a control algorithm running on the real hardware.

The only oddity in my mind is that there doesn't seem to be a SpiceSharp Netlist Parser that's kept up-to-date with SpiceSharp itself.  Therefore, I wrote a parser.  After all, they are pretty easy to parse. Am I just missing something obvious though?

**Reply by svenboulanger** [reply]
Happy to hear you're liking it!

We are maintaining a parser [in this repository](https://github.com/SpiceSharp/SpiceSharpParser). I am keeping in contact with @marcin-golebiowski , who is main author and contributor of Spice#.Parser. He is still working on the project and regularly notifies me of bugs and features that he thinks should be included, but since he follows the Spice# updates it will not be simultaneous and depending on the situation it can take a while, just like it sometimes can take a while for me to make a new update of Spice# :-). I can confirm that he's still working on it though, but we have to combine these projects with our daily lives so we can't guarantee that it will always be swift.

If you stick to very basic Spice-syntax, then it is indeed relatively straight-forward to parse, however I know from Marcin's work that supporting many different Spice-flavors (TSpice, HSpice, LTSpice, ...) is a whole different endeavor.

**Comment by DeadTomGC** [comment]
I started off using that parser, but it didn't support the features I needed most for the work I was doing since it wasn't quite up-to-date.  but I still appreciate all the work that's gone into all 4 of the associated projects.  I've learned a lot from all of them.
